Erroneous type error with create-react-app and `noUncheckedIndexedAccess`

**Describe the bug**
If you use the `noUncheckedIndexedAccess` compiler flag, classnames that are defined will still be typed as `string | undefined`

**To Reproduce**
Repro repo: https://github.com/atomanyih/ts-css-modules-test

If you `npm start` or `npx tsc` you will see this type error:

```
error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.
  Type 'undefined' is not assignable to type 'string'.

12   const className = someFunction(styles.test);
                                    ~~~~~~~~~~~
```

**Expected behavior**
The generated classname should be defined

**Additional context**
Use-cases where this is annoying:
- using `classnames`
`cx({[styles.someClass]: someCondition})` will error.
Workaround: `cx(someCondition && styles.someClass)` will pass
- external packages that don't accept undefined
```ts
type someFunction = (className: string) => void

someFunction(styles.someClass) // will error
```
Workaround: `someFunction(styles.someClass as string)` (kind of whack)

I'm migrating a codebase to use the `noUncheckedIndexedAccess` flag and this issue is widely present in the codebase. I am able to work around it, but it's not super ideal

--- a/README.md
+++ b/README.md
@@ -97,17 +97,18 @@ const b = styles['my_other-class'];
 
 Please note that no options are required. However, depending on your configuration, you may need to customise these options.
 
-| Option               | Default value                      | Description                                                                                            |
-| -------------------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------ |
-| `classnameTransform` | `asIs`                             | See [`classnameTransform`](#classnameTransform) below.                                                 |
-| `customMatcher`      | `"\\.module\\.(c\|le\|sa\|sc)ss$"` | Changes the file extensions that this plugin processes.                                                |
-| `customRenderer`     | `false`                            | See [`customRenderer`](#customRenderer) below.                                                         |
-| `customTemplate`     | `false`                            | See [`customTemplate`](#customTemplate) below.                                                         |
-| `goToDefinition`     | `false`                            | Enables jump to definition, with limited compatibility. See [`goToDefinition`](#goToDefinition) below. |
-| `namedExports`       | `true`                             | Enables named exports for compatible classnames.                                                       |
-| `dotenvOptions`      | `{}`                               | Provides options for [`dotenv`](https://github.com/motdotla/dotenv#options).                           |
-| `postcssOptions`     | `{}`                               | See [`postcssOptions`](#postcssOptions) below.                                                         |
-| `rendererOptions`    | `{}`                               | See [`rendererOptions`](#rendererOptions) below.                                                       |
+| Option                     | Default value                      | Description                                                                                            |
+| -------------------------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------ |
+| `classnameTransform`       | `asIs`                             | See [`classnameTransform`](#classnameTransform) below.                                                 |
+| `customMatcher`            | `"\\.module\\.(c\|le\|sa\|sc)ss$"` | Changes the file extensions that this plugin processes.                                                |
+| `customRenderer`           | `false`                            | See [`customRenderer`](#customRenderer) below.                                                         |
+| `customTemplate`           | `false`                            | See [`customTemplate`](#customTemplate) below.                                                         |
+| `goToDefinition`           | `false`                            | Enables jump to definition, with limited compatibility. See [`goToDefinition`](#goToDefinition) below. |
+| `noUncheckedIndexedAccess` | `false`                            | Enable for compatibility with TypeScript's `noUncheckedIndexedAccess`.                                 |
+| `namedExports`             | `true`                             | Enables named exports for compatible classnames.                                                       |
+| `dotenvOptions`            | `{}`                               | Provides options for [`dotenv`](https://github.com/motdotla/dotenv#options).                           |
+| `postcssOptions`           | `{}`                               | See [`postcssOptions`](#postcssOptions) below.                                                         |
+| `rendererOptions`          | `{}`                               | See [`rendererOptions`](#rendererOptions) below.                                                       |
 
 ```json
 {
@@ -195,7 +196,7 @@ The `classes` object represents all the classnames extracted from the CSS Module
 
 #### `goToDefinition`
 
-This allows an editor like Visual Studio Code to jump to a classname's definition (file and line).
+This allows an editor like Visual Studio Code to go to a classname's definition (file and line).
 
 This is experimental, and only works with Sass (for now) and may not always work as expected.
 

--- a/src/helpers/__tests__/getDtsSnapshot.test.ts
+++ b/src/helpers/__tests__/getDtsSnapshot.test.ts
@@ -253,4 +253,33 @@ describe('utils / cssSnapshots', () => {
       expect(dts).toMatchSnapshot();
     });
   });
+
+  describe('with noUncheckedIndexedAccess enabled', () => {
+    const fileName = join(__dirname, 'fixtures', 'test.module.scss');
+    const css = readFileSync(fileName, 'utf8');
+    const options: Options = {
+      classnameTransform: 'camelCaseOnly',
+      noUncheckedIndexedAccess: true,
+    };
+
+    const cssExports = getCssExports({
+      css,
+      fileName,
+      logger,
+      options,
+      processor,
+      compilerOptions,
+    });
+
+    it('should return a dts file with only possibly undefined strings', () => {
+      const dts = createDtsExports({
+        cssExports,
+        fileName,
+        logger,
+        options,
+      });
+      expect(dts).not.toMatch(/\w'?: string/);
+      expect(dts).toMatchSnapshot();
+    });
+  });
 });

--- a/src/helpers/classTransforms.ts
+++ b/src/helpers/classTransforms.ts
@@ -2,7 +2,7 @@ import camelCase from 'lodash.camelcase';
 import { ClassnameTransformOptions } from '../options';
 
 // The below is based on the CSS Modules implementation found here:
-// https://github.com/webpack-contrib/css-loader/blob/master/lib/compile-exports.js
+// https://github.com/webpack-contrib/css-loader
 
 const dashCase = (className: string): string =>
   className.replace(/-+(\w)/g, (_match: string, firstLetter: string) =>

--- a/src/helpers/createDtsExports.ts
+++ b/src/helpers/createDtsExports.ts
@@ -8,10 +8,6 @@ import { Logger } from './logger';
 const isValidVariable = (className: string) =>
   VALID_VARIABLE_REGEXP.test(className);
 
-const classNameToProperty = (className: string) => `'${className}': string;`;
-const classNameToNamedExport = (className: string) =>
-  `export let ${className}: string;`;
-
 const flattenClassNames = (
   previousValue: string[] = [],
   currentValue: string[],
@@ -29,6 +25,13 @@ export const createDtsExports = ({
   options: Options;
 }): string => {
   const classes = cssExports.classes;
+
+  const possiblyUndefined = Boolean(options.noUncheckedIndexedAccess);
+
+  const classNameToProperty = (className: string) =>
+    `'${className}'${possiblyUndefined ? '?' : ''}: string;`;
+  const classNameToNamedExport = (className: string) =>
+    `export let ${className}${possiblyUndefined ? '?' : ''}: string;`;
 
   const processedClasses = Object.keys(classes)
     .map(transformClasses(options.classnameTransform))

--- a/src/index.ts
+++ b/src/index.ts
@@ -32,9 +32,8 @@ function init({ typescript: ts }: { typescript: typeof tsModule }) {
     process.chdir(directory);
 
     // User options for plugin.
-
-    const config = info.config as { options?: Options };
-    const options = config.options ?? {};
+    const options: Options =
+      (info.config as { options?: Options }).options ?? {};
     logger.log(`options: ${JSON.stringify(options)}`);
 
     // Load environment variables like SASS_PATH.

--- a/src/options.ts
+++ b/src/options.ts
@@ -27,8 +27,9 @@ export interface Options {
   dotenvOptions?: DotenvConfigOptions;
   goToDefinition?: boolean;
   namedExports?: boolean;
+  noUncheckedIndexedAccess?: boolean;
   postcssOptions?: PostcssOptions;
-  /** @deprecated To align with other projects. */
+  /** @deprecated To align with naming in other projects. */
   postCssOptions?: PostcssOptions;
   rendererOptions?: RendererOptions;
 }

